src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -259,83 +259,121 @@ private JsonPropertyInfo GetPropertyWithUniqueAttribute(Type attributeType, Dict
+                            {
+                                if (propertyName.Length <= PropertyNameKeyLength ||
+                                    propertyName.SequenceEqual(propertyRef.Info.Name))
+                                {
+                                    return propertyRef.Info;
+                                }
+                            }
+
+                            --iBackward;
+                        }
+                    }
+                    else if (iBackward >= 0)
                     {
-                        if (TryIsPropertyRefEqual(localPropertyRefsSorted[iBackward], propertyName, key, ref info))
+                        // This path is the same code as above; copied here instead of creating a method since this is a hot path.
+                        PropertyRef propertyRef = localPropertyRefsSorted[iBackward];
+                        if (key == propertyRef.Key)
                         {
-                            return info;
+                            if (propertyName.Length <= PropertyNameKeyLength ||
+                                propertyName.SequenceEqual(propertyRef.Info.Name))
+                            {
+                                return propertyRef.Info;
+                            }
                         }
+
                         --iBackward;
                     }
+                    else
+                    {
+                        // Property was not found.
+                        break;
+                    }
                 }
             }
 
             // No cached item was found. Try the main list which has all of the properties.
 
             string stringPropertyName = JsonHelpers.Utf8GetString(propertyName);
-            if (PropertyCache.TryGetValue(stringPropertyName, out info))
+            if (!PropertyCache.TryGetValue(stringPropertyName, out JsonPropertyInfo info))
             {
-                // Check if we should add this to the cache.
-                // Only cache up to a threshold length and then just use the dictionary when an item is not found in the cache.
-                int count;
-                if (localPropertyRefsSorted != null)
-                {
-                    count = localPropertyRefsSorted.Length;
-                }
-                else
+                info = JsonPropertyInfo.s_missingProperty;
+            }
+
+            // Three code paths to get here:
+            // 1) info == s_missingProperty. Property not found.
+            // 2) key == info.PropertyNameKey. Exact match found.
+            // 3) key != info.PropertyNameKey. Match found due to case insensitivity.
+            Debug.Assert(info == null || key == info.PropertyNameKey || Options.PropertyNameCaseInsensitive);
+
+            // Check if we should add this to the cache.
+            // Only cache up to a threshold length and then just use the dictionary when an item is not found in the cache.
+            int cacheCount;
```

Comment:
nit:
```C#
int cacheCount = 0;
if (localPropertyRefsSorted != null)
{
    cacheCount = localPropertyRefsSorted.Length;
}
```

Reply:
Is that any faster? That would initialize to 0 first and then overwrite it.